h2o-algos/src/main/java/hex/util/LinearAlgebraUtils.java

```diff
@@ -164,23 +166,38 @@ public BMulInPlaceTask(DataInfo xinfo, double[][] yt, int nColsExp) {
 
     public double[][] _atq;    // Output: A'Q is p_exp by k, where p_exp = number of cols in A with categoricals expanded
 
-    public SMulTask(DataInfo ainfo, int ncolQ) {
+    public SMulTask(DataInfo ainfo, int ncolQ, int ncolExp) {
       _ainfo = ainfo;
       _ncolA = ainfo._adaptedFrame.numCols();
-      _ncolExp = numColsExp(ainfo._adaptedFrame,true);
+      _ncolExp = ncolExp;   // when call from GLRM or PCA
       _ncolQ = ncolQ;
+      _atq = new double[_ncolExp][_ncolQ];
     }
 
-    public SMulTask(DataInfo ainfo, int ncolQ, int ncolExp) {
+    public SMulTask(DataInfo ainfo, int ncolQ) {
+      this(ainfo, ncolQ, numColsExp(ainfo._adaptedFrame,true));
+    }
+
+    public SMulTask(DataInfo ainfo, int ncolQ, int ncolExp, double[][] atq) {
       _ainfo = ainfo;
       _ncolA = ainfo._adaptedFrame.numCols();
       _ncolExp = ncolExp;   // when call from GLRM or PCA
       _ncolQ = ncolQ;
+      _atq = atq; // zero out the entries before proceeding
+      int atqLen = atq.length;
+      assert ((atqLen==_ncolExp) && (atq[0].length==_ncolQ)); // make sure we do not have null arrays
+      for (int index = 0; index < atqLen; index++) {
+        Arrays.fill(_atq[index], 0);
+      }
+    }
+
+    // avoid memory allocation with this one
+    public SMulTask(DataInfo ainfo, int ncolQ, double[][] atq) {
+      this(ainfo, ncolQ, numColsExp(ainfo._adaptedFrame,true), atq);
     }
 
     @Override public void map(Chunk cs[]) {
       assert (_ncolA + _ncolQ) == cs.length;
-      _atq = new double[_ncolExp][_ncolQ];
```

Comment:
See my comment above. The reason for allocating inside map is to have private memory copy we can modify freely. It looks like _atq is modified inside the loop, so it is a bug unless racy behavior is expected.